lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
+    if (!userObject && !(extraQueryParameters && (extraQueryParameters.indexOf(Constants.login_hint) !== -1 ))) {
+    // TODO: This should be replaced with error callback
+    // TODO: Is this always accessToken?
+      if (this._tokenReceivedCallback) {
+        this._logger.info("User login is required");
+        this._tokenReceivedCallback(ErrorDescription.userLoginError, null, ErrorCodes.userLoginError, Constants.accessToken, this.getUserState(this._cacheStorage.getItem(Constants.stateLogin, this.storeAuthStateInCookie)));
+        return;
+      }
+    }
+
+    // Track the acquireToken progress
+    this._acquireTokenInProgress = true;
+
+    let authenticationRequest: AuthenticationRequestParameters;
+    let acquireTokenAuthority = authority ? AuthorityFactory.CreateInstance(authority, this.validateAuthority) : this.authorityInstance;
+
+    // TODO: Set response type here
+    acquireTokenAuthority.ResolveEndpointsAsync().then(() => {
+      // On Fulfillment
+      if (Utils.compareObjects(userObject, this.getUser())) {
+        if (scopes.indexOf(this.clientId) > -1) {
+          authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.id_token, this.getRedirectUri(), this._state);
+        } else {
+          authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.token, this.getRedirectUri(), this._state);
+        }
+      } else {
+        authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.id_token_token, this.getRedirectUri(), this._state);
+      }
+
+      // Cache nonce
+      this._cacheStorage.setItem(Constants.nonceIdToken, authenticationRequest.nonce, this.storeAuthStateInCookie);
+
+      // Cache acquireTokenUserKey
+      var acquireTokenUserKey;
+      if (userObject) {
+        acquireTokenUserKey = Constants.acquireTokenUser + Constants.resourceDelimeter + userObject.userIdentifier + Constants.resourceDelimeter + authenticationRequest.state;
+      } else {
+        acquireTokenUserKey = Constants.acquireTokenUser + Constants.resourceDelimeter  + Constants.no_user + Constants.resourceDelimeter + authenticationRequest.state;
+      }
+      this._cacheStorage.setItem(acquireTokenUserKey, JSON.stringify(userObject));
+
+      // Cache authorityKey
+      const authorityKey = Constants.authority + Constants.resourceDelimeter + authenticationRequest.state;
+      this._cacheStorage.setItem(authorityKey, acquireTokenAuthority.CanonicalAuthority, this.storeAuthStateInCookie);
+
+      // Set extraQueryParameters to be sent to the server
+      if (extraQueryParameters) {
+        authenticationRequest.extraQueryParameters = extraQueryParameters;
+      }
+
+      // Construct urlNavigate
+      let urlNavigate = authenticationRequest.createNavigateUrl(scopes)   + Constants.response_mode_fragment;
```

Comment:
to do, get consistency on usage of var, let, and const

Reply:
Will be part of the refactoring, most likely in the Request or Response PRs.